README.md
# Scrolli

## What is it?

Scrolli is a jQuery plugin for creating scrolled interfaces easily, straight to the point. Need a scroller? Add Scrolli to your web page, write some markup… and it's done!

Need more control? Don't worry. Scrolli comes with a set of tools to extend it's functionality and adapt it to your needs.

## Setup

Download the [latest copy of Scrolli](https://github.com/juanghurtado/jquery-scrolli/archives/master) and reference it on your `HEAD` element:

``` html
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.6/jquery.min.js"></script>
<script src="jquery.scrolli.js"></script>
<script>
  jQuery(function() {
    jQuery('.scrolli').scrolli();
  });
</script>
```

Add the Scrolli CSS default theme to spice it up visually. You don't need to do this if you want to customize Scrolli by yourself:

``` html
<link rel="stylesheet" href="css/jquery.scrolli.default.css">
```

Build your scroller with this HTML structure, writing anything you want inside the `.scrolli-element` blocks:

``` html
<div class="scrolli">
  <div class="scrolli-element">
    <!-- Content for the scroller (Can be anything) -->
    <img src="images/image-01.jpg" />
  </div><!-- .scrolli-element -->

  <div class="scrolli-element">
    <!-- Content for the scroller (Can be anything) -->
    <img src="images/image-01.jpg" />
    <p>Lorem ipsum dolor sit amet</p>
  </div><!-- .scrolli-element -->

  <div class="scrolli-element">
    <!-- Content for the scroller (Can be anything) -->
    <img src="images/image-03.jpg" />
  </div><!-- .scrolli-element -->

  <div class="scrolli-element">
    <!-- Content for the scroller (Can be anything) -->
    <img src="images/image-04.jpg" />
  </div><!-- .scrolli-element -->
</div><!-- .scrolli -->
```

## Documentation

### Configuration parameters

Scrolli gives you a set of configuration vars to customize its behaviour. You can pass this vars to the Scrolli call like this:

``` javascript
jQuery(function() {
  jQuery('.scrolli').scrolli({
    configVar1 : 'value',
    configVar3 : 'value',
    configVar3 : 'value',
  });
});
```

This are the available configuration vars:

* **showArrowPager** - (Default: `true`) Boolean value that determinates if the arrow pager ('Move left' and 'Move right') should be shown
* **showNumericPager** - (Default: `true`) Boolean value that determinates if the numeric pager should be shown
* **pagerLeftLabel** - (Default: `"Move left"`) Label for the arrow pager item that moves the scroller to the left
* **pagerRightLabel** - (Default: `"Move right"`) Label for the arrow pager item that moves the scroller to the right
* **autoplay** - (Default: `false`) Boolean value that determinates if the scroller automatically move
* **autoplayDelay** - (Default: `10000`) Numeric value (milliseconds) that determinates the amount of time between every automatic move from the autoplay
* **moveSpeed** - (Default: `200`) Numeric value (milliseconds) that determinates the speed of the move animation
* **moveEasing** - (Default: `"linear"`) Easing for the move animation
* **beforeMove** - Method executed before the move animation. Recieves the target element as a parameter
* **afterMove** - Method executed after the move animation. Recieves the target element as a parameter

### Public methods

There are also some public methods available if you want to control Scrolli from Javascript.

* **move_to(target)** - Moves the scroller to the position of the element recieved as a parameter
* **move\_to\_prev()** - Moves the scroller to the previous element on the list
* **move\_to\_next()** - Moves the scroller to the next element on the list

You can call this methods through a `scrolli' data value attached to the Scrolli main element:

``` javascript
$('.scrolli').data('scrolli').move_to($('.scrolli .scrolli-element').last());
$('.scrolli').data('scrolli').move_to_prev();
$('.scrolli').data('scrolli').move_to_next();
```

## Contribute

If you want to contribute in the development of Scrolli, please [fork this project](https://github.com/juanghurtado/jquery-scrolli), add your functionality and send me a pull request. If needed, be sure to add some [Jasmine](http://pivotal.github.com/jasmine/ "Jasmine: BDD for your JavaScript") tests under `/tests/spec/scrolli.js` to make sure everything is ok.

## Credits

* Juan G. Hurtado [[hello@juanghurtado.com](mailto:&#x63;&#x72;&#x65;&#x61;&#x74;&#x69;&#x76;&#x6F;&#x40;&#x61;&#x6C;&#x76;&#x61;&#x72;&#x6F;&#x67;&#x72;&#x61;&#x66;&#x69;&#x63;&#x6F;&#x2E;&#x65;&#x73;)]
* Javier Rico [[jrico@zuinqstudio.com](mailto:&#x63;&#x72;&#x65;&#x61;&#x74;&#x69;&#x76;&#x6F;&#x40;&#x61;&#x6C;&#x76;&#x61;&#x72;&#x6F;&#x67;&#x72;&#x61;&#x66;&#x69;&#x63;&#x6F;&#x2E;&#x65;&#x73;)]
* Álvaro Fernández [[creativo@alvarografico.es](mailto:&#x63;&#x72;&#x65;&#x61;&#x74;&#x69;&#x76;&#x6F;&#x40;&#x61;&#x6C;&#x76;&#x61;&#x72;&#x6F;&#x67;&#x72;&#x61;&#x66;&#x69;&#x63;&#x6F;&#x2E;&#x65;&#x73;)]


CHANGELOG.md
# Changelog

## v0.1 (August 12, 2011)

* First functional version

tests/spec/scrolli.js
describe("Scrolli", function() {
  beforeEach(function() {
    scrolli         = '.scrolli',
    wrapper         = '.scrolli-wrapper',
    elementWrappers = '.scrolli .scrolli-element-wrapper',
    currentElement  = '.scrolli .scrolli-element-wrapper.current',
    pagerLeft       = '.scrolli-wrapper .scrolli-pager-left',
    pagerRight      = '.scrolli-wrapper .scrolli-pager-right',
    pagerItem       = '.scrolli-wrapper .scrolli-pager .scrolli-pager-item';
  });

  describe("Global behaviour", function() {
    beforeEach(function() {
      loadFixtures('four-items.html');
    });

    it('should be marked as processed', function () {
      $(scrolli).scrolli();
      expect($(scrolli)).toHaveClass('scrolli-processed');
    });

    it('should show first item on load', function () {
      $(scrolli).scrolli();
      expect($(elementWrappers).first()).toHaveClass('current');
    });
  });

  describe("Move behaviour", function() {
    beforeEach(function() {
      loadFixtures('four-items.html');
    });

    it('should not move to prev if current item is the first one', function () {
      $(scrolli).scrolli({
        moveSpeed : 0 // Remove animation speed to provoke an instant move
      });
      $(scrolli).data('scrolli').move_to_prev();

      expect($(scrolli).css('left')).toBe('0px');
    });

    it('should not move to next if current item is the last one', function () {
      $(scrolli).scrolli({
        moveSpeed : 0 // Remove animation speed to provoke an instant move
      });
      $(scrolli).data('scrolli').move_to($(elementWrappers).last());
      $(scrolli).data('scrolli').move_to_next();
      
      // element width * last item index
      var left = $(elementWrappers).first().width() * ($(elementWrappers).last().index());

      expect($(scrolli).css('left')).toBe((left * -1) +'px');
    });

    it('should move to prev if current item is not the first one', function () {
      $(scrolli).scrolli({
        moveSpeed : 0 // Remove animation speed to provoke an instant move
      });
      
      $element = $(elementWrappers).last().prev();
      
      $(scrolli).data('scrolli').move_to($element);
      $(scrolli).data('scrolli').move_to_prev();
      
      // element width * last item index
      var left = $element.width() * ($element.prev().index());

      expect($(scrolli).css('left')).toBe((left * -1) +'px');
    });

    it('should move to next if current item is not the last one', function () {
      $(scrolli).scrolli({
        moveSpeed : 0 // Remove animation speed to provoke an instant move
      });
      
      $element = $(elementWrappers).first();
      
      $(scrolli).data('scrolli').move_to($element);
      $(scrolli).data('scrolli').move_to_next();
      
      // element width * last item index
      var left = $element.width() * ($element.next().index());

      expect($(scrolli).css('left')).toBe((left * -1) +'px');
    });
  });

  describe("Autoplay behaviour", function() {

  });

  describe("Arrow pager behaviour", function() {
    describe("With one item", function() {
      beforeEach(function() {
        loadFixtures('one-item.html');
      });
      
      it('should hide arrow pager', function () {
        $(scrolli).scrolli();
        expect($(wrapper)).not.toContain('.scrolli-pager-left');
        expect($(wrapper)).not.toContain('.scrolli-pager-right');
      });
    });
    
    describe("With four items", function() {
      beforeEach(function() {
        loadFixtures('four-items.html');
      });

      it('should show arrow pager by default', function () {
        $(scrolli).scrolli();
        expect($(wrapper)).toContain('.scrolli-pager-left');
        expect($(wrapper)).toContain('.scrolli-pager-right');
      });

      it('should hide arrow pager through config option', function () {
        $(scrolli).scrolli({
          showArrowPager : false
        });
        expect($(wrapper)).not.toContain('.scrolli-pager-left');
        expect($(wrapper)).not.toContain('.scrolli-pager-right');
      });

      it('should hide pager-right link if current item is the first one', function () {
        $(scrolli).scrolli();
        expect($(pagerRight)).not.toBeVisible();
      });

      it('should hide pager-left link if current item is the last one', function () {
        $(scrolli).scrolli();
        $(scrolli).data('scrolli').move_to($(elementWrappers).last());
        expect($(pagerLeft)).not.toBeVisible();
      });
      
      it('should move to the next item when clicking on pager-left', function () {
        $(scrolli).scrolli({
          moveSpeed : 0 // Remove animation speed to provoke an instant move
        });

        $(pagerLeft).click();
        
        $currentElement = $(currentElement);

        // element width * last item index
        var left = $currentElement.width() * ($currentElement.index());

        expect($(scrolli).css('left')).toBe((left * -1) +'px');
      });
      
      it('should move to the prev item when clicking on pager-right', function () {
        $(scrolli).scrolli({
          moveSpeed : 0 // Remove animation speed to provoke an instant move
        });
        
        $(scrolli).data('scrolli').move_to($(elementWrappers).last());
        $(pagerRight).click();
        
        $currentElement = $(currentElement);

        // element width * last item index
        var left = $currentElement.width() * ($currentElement.index());

        expect($(scrolli).css('left')).toBe((left * -1) +'px');
      });
      
      it('should not move to the next item when clicking on pager-left if we are on last item', function () {
        $(scrolli).scrolli({
          moveSpeed : 0 // Remove animation speed to provoke an instant move
        });
        
        $(scrolli).data('scrolli').move_to($(elementWrappers).last());
        $(pagerLeft).click();

        // element width * last item index
        var left = $(elementWrappers).last().width() * ($(elementWrappers).last().index());

        expect($(scrolli).css('left')).toBe((left * -1) +'px');
      });
      
      it('should not move to the prev item when clicking on pager-right if we are on first item', function () {
        $(scrolli).scrolli({
          moveSpeed : 0 // Remove animation speed to provoke an instant move
        });
        
        $(pagerRight).click();

        // element width * last item index
        var left = $(elementWrappers).first().width() * ($(elementWrappers).first().index());

        expect($(scrolli).css('left')).toBe((left * -1) +'px');
      });
    });
  });
  
  describe("Numeric pager behaviour", function() {
    describe("With one item", function() {
      beforeEach(function() {
        loadFixtures('one-item.html');
      });
      
      it('should hide numeric pager', function () {
        $(scrolli).scrolli();
        expect($(wrapper)).not.toContain('.scrolli-pager');
      });
    });
    
    describe("With four items", function() {
      beforeEach(function() {
        loadFixtures('four-items.html');
      });

      it('should show numeric pager by default', function () {
        $(scrolli).scrolli();
        expect($(wrapper)).toContain('.scrolli-pager');
      });

      it('should hide numeric pager through config option', function () {
        $(scrolli).scrolli({
          showNumericPager : false
        });
        expect($(wrapper)).not.toContain('.scrolli-pager');
      });

      it('should mark current item correctly on numeric pager on load', function() {
        $(scrolli).scrolli();
        expect($(pagerItem).first()).toHaveClass('current');

        $(scrolli).data('scrolli').move_to($(elementWrappers).last());

        expect($(pagerItem).first()).not.toHaveClass('current');
        expect($(pagerItem).last()).toHaveClass('current');
      });

      it('should mark current item correctly on numeric pager after using move_to method', function() {
        $(scrolli).scrolli();
        $(scrolli).data('scrolli').move_to($(elementWrappers).last());

        expect($(pagerItem).last()).toHaveClass('current');
      });

      it('should mark current item correctly on numeric pager after using move_to_next method', function() {
        $(scrolli).scrolli();
        $(scrolli).data('scrolli').move_to_next();

        expect($(pagerItem).first()).not.toHaveClass('current');
        expect($(pagerItem).first().next()).toHaveClass('current');
      });

      it('should mark current item correctly on numeric pager after using move_to_prev method', function() {
        $(scrolli).scrolli({
          moveSpeed : 0 // Remove animation speed to provoke an instant move
        });
        $(scrolli).data('scrolli').move_to($(elementWrappers).last());
        $(scrolli).data('scrolli').move_to_prev();

        expect($(pagerItem).first()).not.toHaveClass('current');
        expect($(pagerItem).last()).not.toHaveClass('current');
        expect($(pagerItem).last().prev()).toHaveClass('current');
      });
    });
  });
});

jquery.scrolli.js
/* ----------------------------------------------------------------------------------
  $.scrolli plugin

  Encoding: UTF-8
  Version:  0.1
  Authors:  
    Juan G. Hurtado   [hello@juanghurtado.com]
----------------------------------------------------------------------------------
  Table of contents
----------------------------------------------------------------------------------
  1.DEFAULT CONFIG
  2.LOCAL VARS
  3.INIT METHOD
  4.PUBLIC METHODS
    4.1.move_to_next()
    4.2.move_to_prev()
    4.3.move_to($target)
  5.PRIVATE METHODS
    5.1.update_visual_state()
    5.2.clear_timer()
  6.INITIALIZATION
  7.PLUGIN BINDING
---------------------------------------------------------------------------------- */
(function($) {
  $.scrolli = function(element, options) {
    
    /* =DEFAULT CONFIG
    ------------------------------------------------------------------------------ */
    var defaults = {
      showArrowPager    : true,
      showNumericPager  : true,
      pagerLeftLabel    : 'Move left',
      pagerRightLabel   : 'Move right',
      autoplay          : false,
      autoplayDelay     : 10000,
      moveSpeed         : 200,
      moveEasing        : 'linear',
      beforeMove        : function($target) {},
      afterMove         : function($target) {}
    }
    
    /* =LOCAL VARS
    ------------------------------------------------------------------------------ */
    var plugin    = this,
        element   = element,
        timer     = null,
        $element  = $(element),
        $scrolli  = $element,
        $wrapper,
        $elements,
        $elementWrappers,
        $currentElement,
        $pagerLeft,
        $pagerRight,
        $numericPager,
        $numericPagerItem;
        
    plugin.settings = {};
    
    /* =INIT METHOD
    ------------------------------------------------------------------------------ */
    plugin.init = function() {
      ps = plugin.settings = $.extend({}, defaults, options);
      
      if ($scrolli.hasClass('scrolli-processed')) {
        return;
      }
        
      // 1.- Build HTML wrappers
      $scrolli
        .wrap('<div class="scrolli-wrapper"></div>')
        .wrap('<div class="scrolli-overflow"></div>')
        .find('.scrolli-element')
        .wrap('<div class="scrolli-element-wrapper"></div>');
           
      // 2.- Define available global vars
      $wrapper          = $scrolli.parent().parent();
      $overflow         = $scrolli.parent();
      $elements         = $scrolli.find('.scrolli-element');
      $elementWrappers  = $scrolli.find('.scrolli-element-wrapper');
      $currentElement   = $elementWrappers.first();
      
      // 3.- Set scroller styling
      $scrolli.css({
        left : '0',
        position : 'relative',
        width : '9999em'
      });
      
      $overflow.css({
        overflow : 'hidden',
        position : 'relative'
      });
      
      $elementWrappers.css({
        'float' : 'left',
        width : $wrapper.width()
      });
        
      // 4.- Build arrow pager system
      if (ps.showArrowPager && $elements.length > 1) {
        $pagerLeft  = $('<a href="#" class="scrolli-pager-left">'+ ps.pagerLeftLabel +'</a>');
        $pagerRight = $('<a href="#" class="scrolli-pager-right">'+ ps.pagerRightLabel +'</a>');

        $pagerLeft.bind('click', function(e) {
          clear_timer();
          plugin.move_to_next();
          e.preventDefault();
        }).appendTo($wrapper);

        $pagerRight.bind('click', function(e) {
          clear_timer();
          plugin.move_to_prev();
          e.preventDefault();
        }).appendTo($wrapper);
      }
      
      // 5.- Build numeric pager system
      if (ps.showNumericPager && $elements.length > 1) {
        $numericPager     = $('<ul class="scrolli-pager"></ul>');
        $numericPagerItem = $('<li class="scrolli-pager-item"><a class="scrolli-pager-link" href="#"></a></li>');
        
        $elements.each(function(index) {
          $numericPagerItem.find('a').text(index + 1);
          $numericPager.append($numericPagerItem.clone());
        });
        
        $numericPager.find('a').bind('click', function(e) {
          plugin.move_to($elementWrappers.eq($(this).parent().index()));
          e.preventDefault();
        });
        
        $numericPager.appendTo($wrapper);
      }
      
      // 6.- Launch autoplay if needed
      if (ps.autoplay) {
        timer = setInterval(function() {
          if ($currentElement.next().length > 0) {
            plugin.move_to_next();
          } else {
            plugin.move_to($elementWrappers.first());
          }
        }, ps.autoplayDelay);
      }
      
      // 7.- Update visual state
      update_visual_state();
      
      // 8.- Mark as processed if everything was fine
      $scrolli.addClass('scrolli-processed');
    }
    
    /* =PUBLIC METHODS
    ------------------------------------------------------------------------------ */
    
    /* =|move_to_next()
     *
     *   Moves the scroller to the next element on the list
    ------------------------------------------------------------------------------ */
    plugin.move_to_next = function() {
      var $prev = $currentElement.next();
      if ($prev.length > 0) {
        plugin.move_to($prev);
      }
    }
    
    /* =|move_to_prev()
     *
     *   Moves the scroller to the previous element on the list
    ------------------------------------------------------------------------------ */
    plugin.move_to_prev = function() {
      var $next = $currentElement.prev();
      if ($next.length > 0) {
        plugin.move_to($next);
      }
    }
    
    /* =|move_to($target)
     *  - $target ($ object): Item to show
     *
     *   Moves the scroller to the target element
    ------------------------------------------------------------------------------ */
    plugin.move_to = function($target) {
      // Cancel if being moved already
      if ($scrolli.is(':animated') || $target == null || $target.length <= 0) {
        return;
      }
      
      // Before move callback
      ps.beforeMove($target);
      
      // Movement amount math
      var target_left   = ($target.index() - $currentElement.index()) * $currentElement.width(),
          current_left  = $scrolli.position().left;
      
      // Animation
      $scrolli.animate({
        left : (current_left - target_left) + 'px'
      },ps.moveSpeed,
        ps.moveEasing,
        ps.afterMove($target));
      
      $currentElement = $target;
      
      update_visual_state();
    }
    
    /* =PRIVATE METHODS
    ------------------------------------------------------------------------------ */
    
    /* =|update_visual_state()
     *
     *   Updates global visual state
    ------------------------------------------------------------------------------ */
    var update_visual_state = function() {
      // Arrow pager visual state
      if (ps.showArrowPager && $elements.length > 1) {
        if ($currentElement.next().length <= 0) {
          $pagerLeft.hide();
        } else {
          $pagerLeft.show();
        }
      
        if ($currentElement.prev().length <= 0) {
          $pagerRight.hide();
        } else {
          $pagerRight.show();
        }
      }
      
      // Numeric pager visual state
      if (ps.showNumericPager && $elements.length > 1) {
        $numericPager.find('.scrolli-pager-item')
        .removeClass('current')
        .eq($currentElement.index()).addClass('current');
      }
      
      // Current element visual state
      $elementWrappers.removeClass('current');
      $currentElement.addClass('current');
    }
    
    /* =|clear_timer()
     *
     *   Clears autoplay timer if exists
    ------------------------------------------------------------------------------ */
    var clear_timer = function() {
      if (ps.autoplay) {
        if (timer != null) {
          clearInterval(timer);
        }
      }
    }
    
    /* =INITIALIZATION
    ------------------------------------------------------------------------------ */
    plugin.init();
  }
  
  /* =PLUGIN BINDING
    ------------------------------------------------------------------------------ */
  $.fn.scrolli = function(options) {
    return this.each(function() {
      if (undefined == $(this).data('scrolli')) {
        var plugin = new $.scrolli(this, options);
        $(this).data('scrolli', plugin);
      }
    });
  }
})($);